### PR TITLE
Add a new HollowUniqueKeyIndex that supports object longevity. Update HollowHashIndex to support object longevity.

### PR DESCRIPTION
**Pull request by Jorel-Master-Of-Scheduling**
Add an alternative to HollowPrimaryKeyIndex called HollowUniqueKeyIndex. It uses HollowDataAccess instead of HollowObjectTypeReadState. This allows HollowUniqueKeyIndex to opt-in to Object Longevity by using the HollowDataAccess from a Consumer rather than the HollowObjectTypeReadState.

HollowObjectTypeReadState is mutable and gets updated with each delta.  If a PrimaryKeyIndex does NOT listen to updates, then it will become stale over time. The internal index structure will remain unchanged, but the HollowPrimaryKeyValueDeriver will start returning incorrect keys as the ordinals get re-used and re-assigned to new records.

By migrating to HollowDataAccess, users can choose to use the HollowObjectTypeReadState (which stays up to date) or use HollowApi.getDataAccess() which may be object-longevity-aware.

### CODE REVIEW

File: hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java [hunk 1]
```diff
@@ -216,8 +217,8 @@ public void endUpdate() {
        reindexHashIndex();
     }
 
-    public HollowReadStateEngine getStateEngine() {
-        return stateEngine;
+    public HollowDataAccess getStateEngine() {
```
Comment (dkoszewnik) [inline hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java]: I don't think we should change the public API here.

File: hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.netflix.hollow.core.index;
+
+import java.util.Collection;
+
+/**
+ * This package is for internal use. Do not depend on it.
```
Comment (dkoszewnik) [inline hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java:6]: Should we mark it deprecated?

**Review by dkoszewnik — APPROVED** [review]
Looks good to me, thank you James.

File: hollow/src/main/java/com/netflix/hollow/core/index/key/PrimaryKey.java [hunk 3]
```diff
@@ -31,6 +33,30 @@
...
+     * @param type             hollow type
+     * @param fieldPaths       field paths for fields that make up the primary key. If no fields are passed in, then create using schema definition of primary key
+     * @return populated primary key
+     */
+    public static PrimaryKey create(HollowDataAccess hollowDataAccess, String type, String... fieldPaths) {
+        if (fieldPaths != null && fieldPaths.length != 0) {
+            return new PrimaryKey(type, fieldPaths);
+        }
+
+        HollowSchema schema = hollowDataAccess.getSchema(type);
```
Comment (mbrilnetflix) [inline hollow/src/main/java/com/netflix/hollow/core/index/key/PrimaryKey.java:53]: nit: can `hollowDataAccess` be null? it's ok if we have fieldPaths

File: hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java (hunk 2 above)
Comment (mbrilnetflix) [reply hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java:6]: +1

**Review by mbrilnetflix — APPROVED** [review]
LGTM

**Review by mbrilnetflix — APPROVED** [review]

File: hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java (hunk 1 above)
Comment (Jorel-Master-Of-Scheduling) [reply hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java]: Oops! Let me fix that.

File: hollow/src/main/java/com/netflix/hollow/core/index/key/PrimaryKey.java (hunk 3 above)
Comment (Jorel-Master-Of-Scheduling) [reply hollow/src/main/java/com/netflix/hollow/core/index/key/PrimaryKey.java:53]: Yes, I think so. Added a comment to the javadoc.

File: hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java (hunk 2 above)
Comment (Jorel-Master-Of-Scheduling) [reply hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java:6]: Done.

**Pull request merged**